### PR TITLE
[FIX] Removing instances of [...]_index_specialisation

### PR DESCRIPTION
**Pull request by eaasna**
This PR is part of [seqan/product_backlog/issues/259](https://github.com/seqan/product_backlog/issues/259)

- Removes concept `fm_index_specialisation` and uses unconstrained types instead
- Removes concept `bi_fm_index_specialisation` and checks for `seqan3::detail::template_specialisation_of<typename mytype::cursor_type, seqan3::bi_fm_index_cursor>` instead
- Removes `concept_check` unit tests for `fm_index_specialisation` and `bi_fm_index_specialisation`

### CODE REVIEW

File: test/unit/search/fm_index/fm_index_collection_test_template.hpp [hunk 1]
```diff
@@ -121,11 +121,11 @@ TYPED_TEST_P(fm_index_collection_test, concept_check)
 {
     using index_t = typename TypeParam::first_type;
 
-    EXPECT_TRUE((seqan3::fm_index_specialisation<index_t>));
+    EXPECT_TRUE((seqan3::detail::template_specialisation_of<index_t, seqan3::fm_index>));
```
Comment (eaasna) [inline test/unit/search/fm_index/fm_index_collection_test_template.hpp:124]: Wasn't sure if this concept_check should just be deleted like in the case of #2339

File: test/unit/search/fm_index/fm_index_collection_test_template.hpp [hunk 2]
```diff
@@ -121,11 +121,11 @@ TYPED_TEST_P(fm_index_collection_test, concept_check)
...
     if constexpr (std::same_as<index_t, seqan3::bi_fm_index<typename index_t::alphabet_type,
                                                             seqan3::text_layout::collection>>)
     {
-        EXPECT_TRUE(seqan3::bi_fm_index_specialisation<index_t>);
+        EXPECT_TRUE(seqan3::detail::template_specialisation_of<index_t, seqan3::bi_fm_index>);
```
Comment (eaasna) [inline test/unit/search/fm_index/fm_index_collection_test_template.hpp:128]: Again, delete or no?

**Comment by marehr** [comment]
```c++
../../../seqan3/include/seqan3/search/fm_index/fm_index_cursor.hpp:530:48: error: ‘text_layout’ has not been declared
  530 |         requires (index_t::text_layout_mode == text_layout::single)
      |                                                ^~~~~~~~~~~
```

An include is missing?

File: include/seqan3/search/fm_index/concept.hpp [hunk 3]
```diff
@@ -69,107 +69,3 @@ SEQAN3_CONCEPT sdsl_index = requires (t sdsl_index)
...
- */
-
-//!\brief The possible text layouts (single, collection) the seqan3::fm_index and seqan3::bi_fm_index can support.
-enum text_layout : bool
-{
-    //!\brief The text is a single range.
-    single,
-    //!\brief The text is a range of ranges.
-    collection
-};
```
Comment (eseiler) [inline include/seqan3/search/fm_index/concept.hpp:87]: This must stay

Comment (eaasna) [reply include/seqan3/search/fm_index/concept.hpp]: Thanks a lot!

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 4]
```diff
@@ -24,6 +24,7 @@
 #include <seqan3/range/concept.hpp>
 #include <seqan3/range/views/drop.hpp>
 
+
```
Comment (MitraDarja) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:27]: Why did you add this new line?

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 5]
```diff
@@ -42,9 +43,9 @@ enum class error_type : uint8_t
 
 /*!\brief The algorithm that performs a unidirectional search on an FM index using trivial backtracking.
  * \tparam configuration_t The search configuration type.
- * \tparam index_t The type of index; must model seqan3::fm_index_specialisation.
+ * \tparam index_t The type of index;
```
Comment (MitraDarja) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:46]:
```suggestion
* \tparam index_t The type of index.
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 6]
```diff
@@ -68,7 +69,7 @@ class unidirectional_search_algorithm : protected policies_t...
 
     /*!\brief Constructs from a configuration object and an index.
      * \tparam configuration_t The search configuration type.
-     * \tparam index_t The type of index; must model seqan3::fm_index_specialisation.
+     * \tparam index_t The type of index;
```
Comment (MitraDarja) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:72]:
```suggestion
* \tparam index_t The type of index.
```

File: include/seqan3/search/fm_index/fm_index_cursor.hpp [hunk 7]
```diff
@@ -66,7 +66,7 @@ struct suffix_array_interval
 };
 
 /*!\brief The SeqAn FM Index Cursor.
- * \tparam index_t The type of the underlying index; must model seqan3::fm_index_specialisation.
+ * \tparam index_t The type of the underlying index;
```
Comment (MitraDarja) [inline include/seqan3/search/fm_index/fm_index_cursor.hpp:69]:
```suggestion
 * \tparam index_t The type of the underlying index.
 ```

**Review by MitraDarja — CHANGES_REQUESTED** [review]
Just some minor things.

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp (hunk 4 above)
Comment (eaasna) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:27]: oops :)

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp (hunk 5 above)
Comment (eaasna) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:46]: yes, thanks for noticing!

**Review by MitraDarja — APPROVED** [review]

File: include/seqan3/search/fm_index/bi_fm_index_cursor.hpp [hunk 8]
```diff
@@ -33,7 +33,8 @@ namespace seqan3
  */
 
 /*!\brief The SeqAn Bidirectional FM Index Cursor.
- * \tparam index_t The type of the underlying index; must model seqan3::bi_fm_index_specialisation.
+ * \tparam index_t The type of the underlying index; index_t::cursor_type must model
+ * seqan3::detail::template_specialisation_of a seqan3::bi_fm_index_cursor.
```
Comment (marehr) [inline include/seqan3/search/fm_index/bi_fm_index_cursor.hpp:37]:
```suggestion
 * \tparam index_t The type of the underlying index; index_t::cursor_type must model
 *                 seqan3::detail::template_specialisation_of a seqan3::bi_fm_index_cursor.
```

File: include/seqan3/search/search.hpp [hunk 9]
```diff
@@ -98,8 +98,7 @@ namespace seqan3
  *
  * \include test/snippet/search/search.cpp
  */
-template <fm_index_specialisation index_t,
-          std::ranges::forward_range queries_t,
+template <typename index_t, std::ranges::forward_range queries_t,
           typename configuration_t = decltype(search_cfg::default_configuration)>
```
Comment (marehr) [inline include/seqan3/search/search.hpp:102]:
```suggestion
template <typename index_t,
          std::ranges::forward_range queries_t,
          typename configuration_t = decltype(search_cfg::default_configuration)>
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 10]
```diff
@@ -28,11 +29,15 @@ namespace seqan3::detail
  * \{
  */
 
-/*!\brief The algorithm that performs a unidirectional search on an FM index using trivial backtracking.
+/*!\brief The algorithm that performs a bidirectional search on a bidirectional FM index using (optimal) search schemes.
  * \tparam configuration_t The search configuration type.
- * \tparam index_t The type of index; must model seqan3::bi_fm_index_specialisation.
+ * \tparam index_t The type of index; index_t::cursor_type must model seqan3::detail::template_specialisation_of
+ * a seqan3::bi_fm_index_cursor.
```
Comment (marehr) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:35]:
```suggestion
 * \tparam index_t The type of index; index_t::cursor_type must model seqan3::detail::template_specialisation_of
 *                 a seqan3::bi_fm_index_cursor.
```

File: include/seqan3/search/search.hpp [hunk 11]
```diff
@@ -190,8 +188,7 @@ inline auto search(queries_t && queries,
 
 // Overload for a single query (not a collection of queries)
 //!\overload
-template <fm_index_specialisation index_t,
-          std::ranges::forward_range query_t,
+template <typename index_t, std::ranges::forward_range query_t,
           typename configuration_t = decltype(search_cfg::default_configuration)>
```
Comment (marehr) [inline include/seqan3/search/search.hpp:192]:
```suggestion
template <typename index_t,
          std::ranges::forward_range query_t,
          typename configuration_t = decltype(search_cfg::default_configuration)>
```

File: include/seqan3/search/search.hpp [hunk 12]
```diff
@@ -170,8 +169,7 @@ inline auto search(queries_t && queries,
 //!\cond DEV
 // Convert query sequence if it does not match the alphabet type of the index.
 //!\overload
-template <fm_index_specialisation index_t,
-          std::ranges::forward_range queries_t,
+template <typename index_t, std::ranges::forward_range queries_t,
           typename configuration_t = decltype(search_cfg::default_configuration)>
```
Comment (marehr) [inline include/seqan3/search/search.hpp:173]:
```suggestion
template <typename index_t,
          std::ranges::forward_range queries_t,
          typename configuration_t = decltype(search_cfg::default_configuration)>
```

File: include/seqan3/search/fm_index/bi_fm_index_cursor.hpp [hunk 13]
```diff
@@ -33,7 +33,8 @@ namespace seqan3
  */
 
 /*!\brief The SeqAn Bidirectional FM Index Cursor.
- * \tparam index_t The type of the underlying index; must model seqan3::bi_fm_index_specialisation.
+ * \tparam index_t The type of the underlying index; index_t::cursor_type must model
+ *                 seqan3::detail::template_specialisation_of a seqan3::bi_fm_index_cursor.
```
Comment (marehr) [inline include/seqan3/search/fm_index/bi_fm_index_cursor.hpp:37]:
```suggestion
 * \tparam index_t The type of the underlying index; This is normally seqan3::bi_fm_index.
```

File: include/seqan3/search/fm_index/fm_index_cursor.hpp [hunk 14]
```diff
@@ -66,7 +66,7 @@ struct suffix_array_interval
 };
 
 /*!\brief The SeqAn FM Index Cursor.
- * \tparam index_t The type of the underlying index; must model seqan3::fm_index_specialisation.
+ * \tparam index_t The type of the underlying index.
```
Comment (marehr) [inline include/seqan3/search/fm_index/fm_index_cursor.hpp:69]:
```suggestion
 * \tparam index_t The type of the underlying index. This is normally seqan3::fm_index.
```

File: test/unit/search/fm_index_cursor/fm_index_cursor_collection_test.cpp [hunk 15]
```diff
@@ -29,7 +29,8 @@ struct fm_index_cursor_collection_test : public ::testing::Test
     using alphabet_type = typename index_type::alphabet_type;
     using text_type = std::vector<alphabet_type>;
 
-    static constexpr bool is_bi_fm_index = seqan3::bi_fm_index_specialisation<index_type>;
+static constexpr bool is_bi_fm_index = seqan3::detail::template_specialisation_of<typename index_type::cursor_type,
+                                                                                  seqan3::bi_fm_index_cursor>;
```
Comment (marehr) [inline test/unit/search/fm_index_cursor/fm_index_cursor_collection_test.cpp:33]:
```suggestion
    static constexpr bool is_bi_fm_index = seqan3::detail::template_specialisation_of<typename index_type::cursor_type,
                                                                                      seqan3::bi_fm_index_cursor>;
```

File: test/unit/search/fm_index_cursor/fm_index_cursor_test.cpp [hunk 16]
```diff
@@ -28,7 +28,8 @@ struct fm_index_cursor_test : public ::testing::Test
     using alphabet_type = typename index_type::alphabet_type;
     using text_type = std::vector<alphabet_type>;
 
-    static constexpr bool is_bi_fm_index = seqan3::bi_fm_index_specialisation<index_type>;
+static constexpr bool is_bi_fm_index = seqan3::detail::template_specialisation_of<typename index_type::cursor_type,
+                                                                                  seqan3::bi_fm_index_cursor>;
```
Comment (marehr) [inline test/unit/search/fm_index_cursor/fm_index_cursor_test.cpp:32]:
```suggestion
    static constexpr bool is_bi_fm_index = seqan3::detail::template_specialisation_of<typename index_type::cursor_type,
                                                                                      seqan3::bi_fm_index_cursor>;
```

**Review by marehr — APPROVED** [review]
LGTM; I just fixed some indentation. Do you want to fix the history of the commits or can this be squashed merged?

**Comment by eaasna** [comment]
@marehr Please squash merge

**Pull request merged**